### ISSUE TITLE
Inconsistent in "EXTENDING PYTORCH" and "save_for_backward"

### ISSUE THREAD

**Issue by li-positive-one** [pytorch/pytorch#81223]
### 📚 The doc issue

In [EXTENDING PYTORCH](https://pytorch.org/docs/stable/notes/extending.html#extending-autograd)

> [save_for_backward()](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.save_for_backward.html#torch.autograd.function.FunctionCtx.save_for_backward) must be used when saving input or output tensors of the forward to be used later in the backward. Anything else, i.e., non-tensors and tensors that are neither input nor output should be stored directly on ctx.

It suggests that the "tensors that are neither input nor output" should be stored directly on ctx.

However in [doc of save_for_backward](https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.save_for_backward.html#torch.autograd.function.FunctionCtx.save_for_backward), "tensors that are neither input nor output" are suggested to be stored by "save_for_backward"

> All tensors intended to be used in the backward pass should be saved with save_for_backward (as opposed to directly on ctx) to prevent incorrect gradients and memory leaks, and enable the application of saved tensor hooks. See [torch.autograd.graph.saved_tensors_hooks](https://pytorch.org/docs/stable/autograd.html#torch.autograd.graph.saved_tensors_hooks).

And the example also show that "tensors that are neither input nor output" should be stored by "save_for_backward"
```
class Func(Function):
    @staticmethod
    def forward(ctx, x: torch.Tensor, y: torch.Tensor, z: int):
        w = x * y * z
        out = x * y + y * z + w
        ctx.save_for_backward(x, y, w, out)
        ctx.z = z  # z is not a tensor
        return out
    @staticmethod
    def backward(ctx, grad_out):
        x, y, w, out = ctx.saved_tensors
        z = ctx.z
        gx = grad_out * (y + y * z)
        gy = grad_out * (x + z + x * z)
        gz = None
        return gx, gy, gz
a = torch.tensor(1., requires_grad=True, dtype=torch.double)
b = torch.tensor(2., requires_grad=True, dtype=torch.double)
c = 4
d = Func.apply(a, b, c)
```

cc @svekars @holly1238 @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
@soulitzer we should update the save_for_backward doc to only save the input/outputs
